<!DOCTYPE html>
<html lang="en">
<head>
        <meta charset="utf-8" />
        <title>Watch Out for Python's HasAttr & AttributeError</title>
        <link href='http://fonts.googleapis.com/css?family=Oswald:300' rel='stylesheet' type='text/css'>
        <link rel="stylesheet" href="http://michaeljones.github.io/blog/theme/css/custom.css" />
        <link href="http://michaeljones.github.io/blog/feeds/all.atom.xml" type="application/atom+xml" rel="alternate" title="michael jones | dev blog Atom Feed" />

            <link rel="stylesheet" href="http://michaeljones.github.io/blog/theme/css/main.css?6a19e19d">

        <!--[if IE]>
            <script src="http://html5shiv.googlecode.com/svn/trunk/html5.js"></script>
        <![endif]-->
</head>

<body id="index" class="home">
        <header id="banner">
                <span id="header-title" class="body">
                    <a href="http://michaeljones.github.io/blog/">michael jones | dev blog </a>
                </span>
                <nav>
                    <ul>
                            <li><a href="http://michaeljones.github.io/blog/category/article.html">article</a></li>
                            <li class="active"><a href="http://michaeljones.github.io/blog/category/blog.html">blog</a></li>
                            <li><a href="http://michaeljones.github.io/blog/category/journal.html">journal</a></li>
                            <li><a href="http://michaeljones.github.io/blog/category/links.html">links</a></li>
                            <li><a href="http://michaeljones.github.io/blog/category/video.html">video</a></li>
                    </ul>
                </nav>
        </header><!-- /#banner -->
        <section>
<section id="content" class="body">
  <article>
    <header>
      <h1 class="entry-title">
        <a href="http://michaeljones.github.io/blog/posts/2014/09/21/watch-out-for-pythons-hasattr-attributeerror/" rel="bookmark"
           title="Permalink to Watch Out for Python's HasAttr & AttributeError">Watch Out for Python's HasAttr &amp; AttributeError</a></h1>
    </header>

    <div class="entry-content">
<footer class="post-info">
        <abbr class="published" title="2014-09-21T00:00:00">
                Sun 21 September 2014
        </abbr>

        <address class="vcard author">
                By <a class="url fn" href="http://michaeljones.github.io/blog/author/michael-jones.html">Michael Jones</a>
        </address>
<p>In <a href="http://michaeljones.github.io/blog/category/blog.html">blog</a>. </p>
<p>tags: <a href="http://michaeljones.github.io/blog/tag/python.html">python</a></p>
</footer><!-- /.post-info -->      <p>I was recently trying to contribute to the <a class="reference external" href="https://github.com/asaglimbeni/django-datetime-widget">django-datetime-widget</a> repository
and encountered an issue with my changes not working but no error being reported
from Django or Python.</p>
<p>I was working on the widget's <tt class="docutils literal">_media</tt> method which is converted to a
read-only <tt class="docutils literal">media</tt> property. Django is meant to use the value of the <tt class="docutils literal">media</tt>
attribute on a widget class to help include any additional CSS, images &amp;
Javascript files into the final web page template that uses the widget.</p>
<p>In order to test for the presence of the <tt class="docutils literal">media</tt> attribute, Django uses the
following code:</p>
<pre class="literal-block">
try:
    base = sup_cls.media
except AttributeError:
    base = Media()
</pre>
<p>It tries to access the attribute and if it receives an <tt class="docutils literal">AttributeError</tt> then
it uses the replacement value <tt class="docutils literal">Media()</tt> instead. This is reasonable for basic
attributes on classes, but when combined with Python properties which can allow
attribute access to result in complex method calls then there is room for
trouble.</p>
<p>My issue was that in my <tt class="docutils literal">_media</tt> implementation I had made a mistake that was
resulting in an <tt class="docutils literal">AttributeError</tt> being raised. And the trouble was that this
was being silently caught by Django which was intepretting it as there not being
a <tt class="docutils literal">media</tt> attribute at all and so I was neither seeing the error or my
expected includes in the final rendered HTML.</p>
<div class="section" id="my-proposed-solution">
<h2>My Proposed Solution</h2>
<p>Hoping to make the world a better place, I headed over to the Django-users
mailing and posted that maybe we could approach this differently. I thought that
maybe the following implementation would avoid this issue:</p>
<pre class="literal-block">
if hasattr(sup_cls, 'media'):
    base = sup_cls.media
else:
    base = Media()
</pre>
<p>I thought that using the <tt class="docutils literal">hasattr</tt> check instead of waiting for an exception
to be raised would avoid this issue. I assumed that all that <tt class="docutils literal">hasattr</tt> did was
to check if the object in question had a particular attribute. Crazy right?</p>
</div>
<div class="section" id="the-problem-with-my-solution">
<h2>The Problem with my Solution</h2>
<p>It turns out this does not work. The kind people on the list educated me in
the ways of <tt class="docutils literal">hasattr</tt> and why this approach doesn't help. I had been foolish
enough not to test my proposed solution and it turns out that the documentation
covers why it won't actually work:</p>
<pre class="literal-block">
hasattr(object, name) -&gt; bool

Return whether the object has an attribute with the given name.
(This is done by calling getattr(object, name) and catching exceptions.)
</pre>
<p>There it is: <tt class="docutils literal">hasattr</tt> is implemented in terms of calling <tt class="docutils literal">getattr</tt> and
catching all exceptions. So it would fail on my buggy implementation of
<tt class="docutils literal">_media</tt> in just the same way that the current test does.</p>
<p>Damn shame really, as it is incredibly confusing to have no results and no
feedback.</p>
<p>Fortunately, Python 3 fixes this issue apparently and Python 2 is stuck with the
current functionality for backwards compatibility.</p>
</div>

    </div><!-- /.entry-content -->

  </article>
</section>
        </section>
        <section id="extras" class="body">
                <div class="social">
                        <h2>social</h2>
                        <ul>
                            <li><a href="http://michaeljones.github.io/blog/feeds/all.atom.xml" type="application/atom+xml" rel="alternate">atom feed</a></li>

                        </ul>
                </div><!-- /.social -->
        </section><!-- /#extras -->

</body>
</html>